File: 2020/day03/main.go
/**
 * day 03 2020
 * https://adventofcode.com/2020/day/3
 *
 * compile:			go build main.go
 * run:				./main < input
 * compile & run:	go run main.go < input
 **/
package main

import (
	"bufio"
	"fmt"
	"os"
)

var hill []string	// the grid...or ski hill
var hillLen int

/*	ride the toboggan down the hill and hit some trees!
 *	...and count them while you're at it. */
 func rideSnowyHill(right int, down int) int {
	x := 0
	y := 0
	treeCount := 0

	for x < hillLen {
		// the width of the grid is 31 so having 'y % 31' brings you
		// to the opposite side if you go past the end
		if (hill[x][y % 31] == '#') {
			treeCount++
		}
		y += right
		x += down
	}
	return treeCount
 }

func main() {
	
	scan := bufio.NewScanner(os.Stdin)
	for scan.Scan() {
		hill = append(hill, scan.Text())
	}

	hillLen = len(hill)

	p1 := rideSnowyHill(3, 1)

	// part 2
	slopes := 1
	slopes *= rideSnowyHill(1, 1);
	slopes *= rideSnowyHill(3, 1);
	slopes *= rideSnowyHill(5, 1);
	slopes *= rideSnowyHill(7, 1);
	slopes *= rideSnowyHill(1, 2);

	fmt.Println("part 1:", p1)
	fmt.Println("part 2:", slopes)
}

File: 2020/day03/main.c
/**
 * day 02 2020
 * https://adventofcode.com/2020/day/2
 *
 * compile:
 *    debug:   gcc main.c -o main -g -fsanitize=address -Wall
 *    release: gcc main.c -o main -O2
 * 
 * run:        ./main < input
 **/
#include <stdio.h>
#include <stdint.h>

#define MAXLINE 35
#define MAXLEN 323	// input is 323 lines long

static uint16_t len = 0;	// variable for file length
static char hill[MAXLEN][MAXLINE];


/*	ride the toboggan down the hill and hit some trees!
 *	...and count them while you're at it. */
uint8_t ride_snowy_hill(uint8_t right, uint8_t down)
{
	uint8_t x = 0; 
	uint8_t y = 0;
	uint8_t tree_count = 0;

	while (x < len)	{
		if (hill[x][y % 31] == '#') tree_count++;
		y += right;
		x += down;
	}
	return tree_count;
}


int main()
{
	int c;
	uint16_t i = 0;
	uint16_t j = 0;

	while ((c = fgetc(stdin)) != EOF) {
		if (c == '\n') {
			i++;
			j = 0;
			len++;
		}
		else hill[i][j++] = (char)c;
	}

	uint32_t slopes = 1;
	slopes *= ride_snowy_hill(1, 1);
	slopes *= ride_snowy_hill(3, 1);
	slopes *= ride_snowy_hill(5, 1);
	slopes *= ride_snowy_hill(7, 1);
	slopes *= ride_snowy_hill(1, 2);

	long p1 = ride_snowy_hill(3, 1);

	printf("part 1: %lu\n", p1);
	printf("part 2: %u\n", slopes);

	return 0;
}
